### PR TITLE
Save 0.6-0.7s of init time by splitting registering lists of signals into its own proc, and optimizing QDELETED

### PR DESCRIPTION
**Pull request by Mothblocks**
- Makes QDELETED use isnull(x) instead of !x, giving about 0.2 to 0.25s of speed.
- Make disposal constructs only update icon state rather than go through expensive overlay code. Unfortunately did not have much effect, but is something they should've been doing nonetheless.
- Makes RegisterSignal only take signals directly as opposed to allocating a fresh list of signals. Very few consumers actually used this and it costs about 0.4s. Also I think this is just a bad API anyway and that separate procs are important

`\bRegisterSignal\((.*)list\(` replaced with `RegisterSignals($1list(`

### CODE REVIEW

**Comment by LemonInTheDark** [comment]
paging @ninjanomnom

**Comment by Mothblocks** [comment]
sure, but i never thought this was a good addition to registersignal in general

**Comment by LemonInTheDark** [comment]
yeah but if it upsets him I'd rather have him upset now, rather then a month from now

**Comment by LemonInTheDark** [comment]
also this should maybe also apply to unregister signal for sanity's sake

**Comment by Mothblocks** [comment]
That would be a separate PR but yeah I agree on a fundamentals level

File: code/datums/components/_component.dm [hunk 1]
```diff
@@ -174,37 +174,49 @@
...
 		return
 
-	var/list/procs = signal_procs
-	if(!procs)
-		signal_procs = procs = list()
-	var/list/target_procs = procs[target] || (procs[target] = list())
-	var/list/lookup = target.comp_lookup
-	if(!lookup)
-		target.comp_lookup = lookup = list()
+	if (islist(signal_type))
```
Comment (LemonInTheDark) [inline code/datums/components/_component.dm:185]: this check has a cost to it. guess ifdefing unit tests here would be poor form

Comment (Mothblocks) [reply code/datums/components/_component.dm:185]: I talked about this before but the cost of this check is pretty small right now. I was thinking we remove this once enough time has passed without the runtime coming through

**Comment by Mothblocks** [comment]
Draft if merged before https://github.com/tgstation/tgstation/pull/71229/files/9b41d322e737d2c75b38a642cdc141fecd0f416f..e78935761681902afd5d40827cb9392bd7b7b56e#diff-618a368ff90b8671288b35a09123432580ee2a1ddb1398b37f7e2bff3fe3ee42R1008-R1013

**Comment by Mothblocks** [comment]
image

yeap

**Comment by Mothblocks** [comment]
@LemonInTheDark you should get on this bc im having to fix conflicts and runtimes every few days

File: code/datums/components/_component.dm [hunk 2]
```diff
@@ -176,37 +176,49 @@
...
+	if(!override && target_procs[signal_type])
+		log_signal("[signal_type] overridden. Use override = TRUE to suppress this warning.\nTarget: [target] ([target.type]) Proc: [proctype]")
+
+	target_procs[signal_type] = proctype
+	var/list/looked_up = lookup[signal_type]
+
+	if(isnull(looked_up)) // Nothing has registered here yet
+		lookup[signal_type] = src
+	else if(looked_up == src) // We already registered here
+		// pass
```
Comment (LemonInTheDark) [inline code/datums/components/_component.dm:212]: return?

File: code/modules/recycling/disposal/construction.dm [hunk 3]
```diff
@@ -40,7 +40,7 @@
 		var/datum/component/simple_rotation/rotcomp = GetComponent(/datum/component/simple_rotation)
 		rotcomp.Rotate(usr, ROTATION_FLIP) // this only gets used by pipes created by RPDs or pipe dispensers
 
-	update_appearance()
+	update_appearance(UPDATE_ICON_STATE)
```
Comment (LemonInTheDark) [inline code/modules/recycling/disposal/construction.dm:43]: does this have emissive blocking issues

Comment (Mothblocks) [reply code/modules/recycling/disposal/construction.dm:43]: dunno but this didnt have any effect anyway

```suggestion
	update_appearance(UPDATE_ICON)
```

File: code/datums/components/_component.dm (hunk 2 above)
Comment (Mothblocks) [reply code/datums/components/_component.dm:212]: alright

**Comment by Mothblocks** [comment]
mother FUCKER

**Comment by Mothblocks** [comment]
@Gamer025

File: code/datums/components/_component.dm (hunk 2 above)
Comment (Mothblocks) [reply code/datums/components/_component.dm:212]:
```suggestion
		return
```

**Pull request merged**